File: flexirule/ruleflow/process/normalization/normalization.py
# Copyright (c) 2025, Bolton and contributors
# For license information, please see license.txt

import re
import unicodedata
from typing import Any

import frappe
from frappe import _

from flexirule.ruleflow.utils.field_resolver import FieldResolver


def execute(context, func=None, config=None):
	"""
	Execute a normalization operation.
	"""
	if not func:
		frappe.throw(_("Operation function name is required"))

	if func not in _OPERATIONS:
		frappe.throw(_("Unknown operation: {0}. Available: {1}").format(func, ", ".join(_OPERATIONS.keys())))

	return _OPERATIONS[func](context, config or {})


# =============================================================================
# TRANSLATION TABLE (borrowed conceptually from UPH)
# =============================================================================

TRANSLATION_TABLE = str.maketrans(
	{
		"أ": "ا",
		"إ": "ا",
		"آ": "ا",
		"ى": "ي",
		"ة": "ه",
		"ؤ": "و",
		"ئ": "ي",
		"ـ": "",
		**{chr(0x660 + i): str(i) for i in range(10)},  # Arabic digits
	}
)

# =============================================================================
# TRANSFORMATIONS (single-arg, stateless)
# =============================================================================

TRANSFORMATIONS: dict[str, callable] = {
	"trim": lambda x: x.strip() if isinstance(x, str) else x,
	"lowercase": lambda x: x.lower() if isinstance(x, str) else x,
	"uppercase": lambda x: x.upper() if isinstance(x, str) else x,
	"casefold": lambda x: x.casefold() if isinstance(x, str) else x,
	"unicode_normalize": lambda x: (unicodedata.normalize("NFKD", x) if isinstance(x, str) else x),
	"translate_chars": lambda x: (x.translate(TRANSLATION_TABLE) if isinstance(x, str) else x),
	"remove_spaces": lambda x: x.replace(" ", "") if isinstance(x, str) else x,
	"remove_extra_spaces": lambda x: (re.sub(r"\s+", " ", x).strip() if isinstance(x, str) else x),
	"remove_punctuation": lambda x: (re.sub(r"[^\w\s]", "", x) if isinstance(x, str) else x),
	"remove_numbers": lambda x: re.sub(r"\d+", "", x) if isinstance(x, str) else x,
	"numeric_only": lambda x: re.sub(r"\D", "", x) if isinstance(x, str) else x,
	"alphanumeric_only": lambda x: re.sub(r"[^\w]", "", x) if isinstance(x, str) else x,
	"slug": lambda x: (
		re.sub(r"[^\w\s-]", "", x).strip().lower().replace(" ", "-") if isinstance(x, str) else x
	),
	"title_case": lambda x: x.title() if isinstance(x, str) else x,
	"name_normalize": lambda x: (" ".join([w.capitalize() for w in x.split()]) if isinstance(x, str) else x),
	"email_normalize": lambda x: (
		f"{x.split('@')[0].split('+')[0].replace('.', '').lower()}@{x.split('@')[1].lower()}"
		if isinstance(x, str) and "@" in x
		else x
	),
	"currency_to_number": lambda x: (re.sub(r"[^\d.]", "", x) if isinstance(x, str) else x),
	"tax_id_clean": lambda x: re.sub(r"[\s-]", "", x) if isinstance(x, str) else x,
	"standard_date": lambda x: (str(frappe.utils.getdate(x)) if isinstance(x, str) and x else x),
}

# =============================================================================
# CORE HELPERS
# =============================================================================


def apply_transformations(value, transformations):
	"""
	Apply a list of transformation keys to a value.
	"""
	if value is None:  # Keep original None handling
		return None

	if not transformations:
		return value

	# Handle list of dicts (Table format from UI)
	# [{ "transformation": "trim" }, ...]
	if isinstance(transformations, list) and transformations and isinstance(transformations[0], dict):
		transformations = [t.get("transformation") for t in transformations if t.get("transformation")]

	# Handle Comma Separated Strings (MultiSelect Tags format)
	if isinstance(transformations, str):
		transformations = [t.strip() for t in transformations.split(",") if t.strip()]

	for (
		transform_name
	) in transformations:  # Renamed 'transform' to 'transform_name' to avoid conflict with TRANSFORMATIONS
		# Backward compat: ignore invalid types
		if not isinstance(transform_name, str):
			continue

		transform_name = transform_name.lower().strip()
		func = TRANSFORMATIONS.get(transform_name)
		if not func:
			frappe.logger().warning(f"Unknown transformation: {transform_name}")  # Keep original warning
			continue

		try:
			value = func(value)  # Apply transformation to 'value' directly
		except Exception as e:  # Keep original error logging
			frappe.log_error(
				title="FlexiRule Normalization Error",
				message=f"Step: {transform_name}\nValue: {value}\n{e}",
			)
	return value


# =============================================================================
# OPERATIONS
# =============================================================================


def normalize_field(context, config):
	"""
	Normalize a field in-place or to a target field.
	"""
	doc = context.get("doc")
	if not doc:
		return None

	source_field = config.get("source_field")
	transformations = config.get("transformations")
	target_field = config.get("target_field")

	# Convert frontend string (newline-separated) into a list
	if isinstance(transformations, str):
		transformations = [t.strip() for t in transformations.split("\n") if t.strip()]

	# Ensure we have a list
	if not isinstance(transformations, list):
		transformations = [transformations] if transformations else []

	value = FieldResolver.resolve_picker(doc, source_field, context)
	if value is None:
		return None

	normalized = apply_transformations(value, transformations)

	# Set to target field (or source field if not specified)
	dest_field = target_field or source_field
	if isinstance(dest_field, list):
		dest_field = dest_field[2]  # fieldname
	doc.set(dest_field, normalized)

	return normalized


def normalize_field_to_context(context, config):
	"""
	Normalize a field value and store in context (vars).
	Does NOT modify the document.
	"""
	doc = context.get("doc")
	if not doc:
		return None

	source_field = config.get("source_field")
	transformations = config.get("transformations")
	context_key = config.get("context_key")

	# Parse transformations if passed as string/JSON
	if isinstance(transformations, str):
		try:
			import json

			transformations = json.loads(transformations)
		except (ValueError, TypeError):
			transformations = [t.strip() for t in transformations.split(",") if t.strip()]

	if not isinstance(transformations, list):
		transformations = [transformations] if transformations else []

	value = FieldResolver.resolve_picker(doc, source_field, context)
	if value is None:
		return None

	normalized = apply_transformations(value, transformations)

	if isinstance(source_field, list):
		source_field_name = source_field[2]
	else:
		source_field_name = source_field

	key = context_key or f"normalized_{source_field_name}"

	if "vars" not in context:
		context["vars"] = {}

	context["vars"][key] = normalized

	return key


def normalize_multiple_fields(context, config):
	"""
	Batch normalize multiple fields with their own transformation configs.
	"""
	doc = context.get("doc")
	if not doc:
		return {}

	field_config = config.get("field_config")
	store_in_context = config.get("store_in_context")

	# Parse field_config if passed as string/JSON
	if isinstance(field_config, str):
		import json

		try:
			field_config = json.loads(field_config)
		except Exception:
			field_config = []

	if not field_config:
		return {}

	results = {}

	for item in field_config:
		fieldname = item.get("fieldname")
		transformations = item.get("transformations", [])
		target_field = item.get("target_field")

		if not fieldname:
			continue

		value = FieldResolver.resolve_picker(doc, fieldname, context)
		if value is None:
			continue

		# Handle transformations string format
		if isinstance(transformations, str):
			transformations = [t.strip() for t in transformations.split("\n") if t.strip()]

		normalized = apply_transformations(value, transformations)

		if store_in_context:
			# Store in context
			key = target_field or f"normalized_{fieldname}"
			if "vars" not in context:
				context["vars"] = {}
			context["vars"][key] = normalized
		else:
			# Set on document
			dest_field = target_field or fieldname
			if isinstance(dest_field, list):
				dest_field = dest_field[2]
			doc.set(dest_field, normalized)

		results[fieldname] = normalized

	return results


# Whitelisted API for testing/preview
@frappe.whitelist()
def preview_normalization(text, transformations):
	"""
	Preview normalization result without saving
	"""
	import json

	if isinstance(transformations, str):
		try:
			transformations = json.loads(transformations)
		except Exception:
			transformations = [t.strip() for t in transformations.split(",")]

	return apply_transformations(text, transformations)


# ============================================================
# DISPATCHER
# ============================================================

_OPERATIONS = {
	"normalize_field": normalize_field,
	"normalize_field_to_context": normalize_field_to_context,
	"normalize_multiple_fields": normalize_multiple_fields,
}

File: flexirule/ruleflow/process/normalization/normalization.js
/**
 * Normalization Process Adapter
 */

// Define Profiles (Syncd with Python)
const PROFILES = {
    "default": [
        "trim", "unicode_normalize", "casefold", "translate_chars", "remove_extra_spaces"
    ],
    "arabic_strict": [
        "trim", "unicode_normalize", "casefold", "translate_chars", "remove_punctuation", "remove_extra_spaces"
    ],
    "email": [
        "trim", "lowercase", "email_normalize"
    ],
    "phone": [
        "numeric_only"
    ],
    "slug": [
        "trim", "lowercase", "remove_punctuation", "remove_extra_spaces", "slug"
    ]
};

// Start Namespace
frappe.provide('flexirule.processes');

flexirule.processes["Normalization"] = {

    meta: {
        version: '1.0',
        title: __('Normalization')
    },

    get_schema(operation_name, context) {
        const operation = this.get_operation(operation_name);
        if (!operation) return null;

        return {
            title: operation.label || operation_name,
            size: 'large',
            fields: typeof operation.get_config_fields === 'function'
                ? operation.get_config_fields(context)
                : []
        };
    },

    get_output_schema(operation_name, config, context) {
        const operation = this.get_operation(operation_name);
        if (operation && typeof operation.get_output_schema === 'function') {
            return operation.get_output_schema(config, context);
        }
        return [];
    },

    get_operation(name) {
        return this.operations.find(op => op.func_name === name);
    },

    get_visible_operations() {
        return this.operations.filter(op => op.visible !== false);
    },

    /**
     * Get default configuration for a specific operation
     */
    get_default_config(operation_name) {
        const operation = this.get_operation(operation_name);
        if (!operation || typeof operation.get_config_fields !== 'function') {
            return {};
        }

        const defaults = {};
        const raw_fields = operation.get_config_fields({});

        raw_fields.forEach(field => {
            // Skip layout fields
            if (['Section Break', 'Column Break', 'HTML'].includes(field.fieldtype)) {
                return;
            }

            // Handle Table fields
            if (field.fieldtype === 'Table') {
                defaults[field.fieldname] = [];
                return;
            }

            // Regular field
            if (field.default !== undefined) {
                defaults[field.fieldname] = field.default;
            }
        });

        return defaults;
    },

    /**
     * Validate configuration
     */
    validate(operation_name, config, context = {}) {
        const operation = this.get_operation(operation_name);
        if (!operation || typeof operation.validate !== 'function') {
            return [];
        }

        const msg = operation.validate(config, context);
        if (msg) {
            return [{ fieldname: '_general', message: msg }];
        }

        return [];
    },

    // Operations Definition
    operations: [
        {
            func_name: "normalize_field",
            label: __("Normalize Field"),
            description: __("Normalize a field in-place or to a target field"),
            icon: "edit",
            color: "#3b82f6",
            get_config_fields: (ctx) => {
                const transformation_field = get_transformation_field_table();
                const profile_field = get_profile_field(true); // Table target

                return [
                    {
                        fieldname: "source_field",
                        label: __("Source Field"),
                        fieldtype: "DocField",
                        options: "vars.document_type",
                        reqd: 1
                    },
                    { fieldtype: "Column Break" },
                    {
                        fieldname: "target_field",
                        label: __("Target Field (Optional)"),
                        fieldtype: "DocField",
                        options: "vars.document_type",
                        description: __("If empty, normalizes in-place.")
                    },
                    {
                        fieldtype: "Section Break",
                        label: __("Transformation Rules")
                    },
                    profile_field,
                    {
                        ...transformation_field,
                        read_only_depends_on: "eval:doc.profile && doc.profile !== ''"
                    },
                ];
            }
        },
        {
            func_name: "normalize_field_to_context",
            label: __("Normalize to Variable"),
            description: __("Normalize a field and store it in context (vars)"),
            icon: "variable",
            color: "#10b981",
            get_config_fields: (ctx) => {
                const transformation_field = get_transformation_field_table();
                const profile_field = get_profile_field(true); // Table target

                return [
                    {
                        fieldname: "source_field",
                        label: __("Source Field"),
                        fieldtype: "DocField",
                        options: "vars.document_type",
                        reqd: 1
                    },
                    { fieldtype: "Column Break" },
                    {
                        fieldname: "context_key",
                        label: __("Variable Name"),
                        fieldtype: "Data",
                        description: __("Key to store in 'vars'. Defaults to normalized_{source_field}")
                    },
                    {
                        fieldtype: "Section Break",
                        label: __("Transformation Rules")
                    },
                    profile_field,
                    {
                        ...transformation_field,
                        read_only_depends_on: "eval:doc.profile && doc.profile !== ''"
                    }
                ];
            },
            get_output_schema: (config, ctx) => {
                const key = config.context_key || (config.source_field ? `normalized_${config.source_field}` : "normalized_value");
                return [
                    {
                        label: key,
                        value: key,
                        type: "Data"
                    }
                ];
            }
        },
        {
            func_name: "normalize_multiple_fields",
            label: __("Batch Normalize"),
            description: __("Batch normalize multiple fields"),
            icon: "list",
            color: "#8b5cf6",
            get_config_fields: (ctx) => {
                const transformation_field = get_transformation_field_multiselect();
                const profile_field = get_profile_field(false); // MultiSelect target

                return [
                    {
                        fieldname: "store_in_context",
                        label: __("Store in Context (Variables)"),
                        fieldtype: "Check",
                        default: 0
                    },
                    {
                        fieldtype: "Section Break",
                        label: __("Field Selection")
                    },
                    {
                        fieldname: "field_config",
                        label: __("Field Configuration"),
                        fieldtype: "Table",
                        reqd: 1,
                        fields: [
                            {
                                fieldname: "fieldname",
                                label: __("Field"),
                                fieldtype: "DocField",
                                options: "vars.document_type",
                                reqd: 1,
                                in_list_view: 1,
                                columns: 3
                            },
                            {
                                ...profile_field,
                                label: __("Profile"),
                                in_list_view: 1,
                                columns: 2
                            },
                            {
                                ...transformation_field,
                                in_list_view: 1,
                                columns: 4,
                                read_only_depends_on: "eval:doc.profile && doc.profile !== ''"
                            },
                            {
                                fieldname: "target_field",
                                label: __("Target Field / Variable"),
                                fieldtype: "Data",
                                in_list_view: 1,
                                columns: 2
                            }
                        ]
                    }
                ];
            },
            get_output_schema: (config, ctx) => {
                if (config.store_in_context) {
                    const fields = config.field_config || [];
                    return fields.map(f => {
                        const key = f.target_field || (f.fieldname ? `normalized_${f.fieldname}` : "var");
                        return { label: key, value: key, type: "Data" };
                    });
                }
                return null;
            }
        }
    ]
};

// Helpers
function get_transformation_field_table() {
    return {
        fieldname: "transformations",
        label: __("Transformations"),
        fieldtype: "MultiCheck",
        columns: 3,
        options: get_transformation_options(),
        reqd: 1
    };
}

function get_transformation_field_multiselect() {
    return {
        fieldname: "transformations",
        label: __("Transformations"),
        fieldtype: "MultiCheck",
        columns: 2,
        options: get_transformation_options(),
        reqd: 1
    };
}

function get_transformation_options() {
    return [
        { label: __("Trim"), value: "trim", description: __("Remove leading and trailing whitespace") },
        { label: __("Lowercase"), value: "lowercase", description: __("Convert all characters to lowercase") },
        { label: __("Uppercase"), value: "uppercase", description: __("Convert all characters to uppercase") },
        { label: __("Casefold"), value: "casefold", description: __("Aggressive lowercase for caseless matching (supports special Unicode characters)") },
        { label: __("Unicode Normalize"), value: "unicode_normalize", description: __("Normalize Unicode characters to NFKD form for consistent representation") },
        { label: __("Translate Chars"), value: "translate_chars", description: __("Translate Arabic/Special characters to their standard base forms") },
        { label: __("Remove Spaces"), value: "remove_spaces", description: __("Remove all spaces from the text") },
        { label: __("Remove Extra Spaces"), value: "remove_extra_spaces", description: __("Replace multiple consecutive spaces with a single space") },
        { label: __("Remove Punctuation"), value: "remove_punctuation", description: __("Remove all punctuation and special characters") },
        { label: __("Remove Numbers"), value: "remove_numbers", description: __("Remove all numeric characters") },
        { label: __("Numeric Only"), value: "numeric_only", description: __("Keep only numeric characters") },
        { label: __("Alphanumeric Only"), value: "alphanumeric_only", description: __("Keep only letters and numbers (removes symbols and spaces)") },
        { label: __("Slug"), value: "slug", description: __("Convert text to a URL-friendly slug (lowercase, alphanumeric, and hyphens)") },
        { label: __("Title Case"), value: "title_case", description: __("Capitalize the first letter of each word") },
        { label: __("Email Normalize"), value: "email_normalize", description: __("Standardize email format (lowercase, remove dots/plus in Gmail addresses)") }
    ];
}

function get_profile_field(is_table_target = true) {
    return {
        fieldname: "profile",
        label: __("Profile"),
        fieldtype: "Select",
        options: [
            { label: "Custom", value: "" },
            { label: "Default", value: "default" },
            { label: "Arabic Strict", value: "arabic_strict" },
            { label: "Email", value: "email" },
            { label: "Phone", value: "phone" },
            { label: "Slug", value: "slug" }
        ],
        onchange: (val, row, ctx) => {
            if (val) {
                // Use spread to copy array and prevent mutation of the PROFILES constant
                const transforms = [...(PROFILES[val] || [])];
                ctx.update_field("transformations", transforms);
            }
        }
    };
}